addon/mixins/adapter-fetch.js

```diff
@@ -0,0 +1,147 @@
+import Ember from 'ember';
+import fetch from 'fetch';
+
+const { assign, RSVP } = Ember;
+
+/**
+ * Helper function that turns the data/body of a request into a query param string.
+ * @param {Object} queryParamsObject
+ * @returns {String}
+ */
+export function serialiazeQueryParams(queryParamsObject) {
+  return Object.keys(queryParamsObject).map((key) => {
+    return `${encodeURIComponent(key)}=${encodeURIComponent(queryParamsObject[key])}`;
+  }).join('&');
+}
+
+/**
+ * Helper function to create a plain object from the response's Headers.
+ * Consumed by the adapter's `handleResponse`.
+ * @param {Headers} headers
+ * @returns {Object}
+ */
+export function headersToObject(headers) {
+  let headerObject = {};
+  headers.forEach((value, key) => {
+    headerObject[key] = value;
+  });
+  return headerObject;
+}
+/**
+ * Helper function that translates the options passed to `jQuery.ajax` into a format that `fetch` expects.
+ * @param {Object} options
+ */
+export function mungOptionsForFetch(options) {
+  const _options = assign({
+    credentials: 'same-origin',
+  }, options);
+
+  _options.method = _options.type;
+
+  // GET and HEAD requests can't have a `body`
+  if (_options.data) {
+    if (_options.method === 'GET' || _options.method === 'HEAD') {
+      _options.url += `?${serialiazeQueryParams(_options.data)}`;
+    } else {
+      _options.body = _options.data;
+    }
+  }
+
+  return _options;
+}
+
+export default Ember.Mixin.create({
+  /**
+   * @param {String} url
+   * @param {String} type
+   * @param {Object} options
+   * @override
+   */
+  ajax(url, type, options) {
+    const requestData = {
+      url,
+      method: type,
+    };
+
+    const hash = this.ajaxOptions(url, type, options);
+
+    return this._ajaxRequest(hash)
+      .then((response) => {
+        if (response.ok) {
+          const bodyPromise = response.json();
+          return this.ajaxSuccess(response, bodyPromise, requestData);
+        }
+        throw this.ajaxError(null, response, requestData);
+      })
+      .catch((error, response, requestData) => {
+        throw this.ajaxError(error, response, requestData);
+      });
+  },
+  /**
+   * Overrides the `_ajaxRequest` method to use `fetch` instead of jQuery.ajax
+   * @param {Object} options
+   * @override
+   */
+  _ajaxRequest(options) {
+    const _options = mungOptionsForFetch(options);
+
+    return fetch(_options.url, _options);
+  },
+
+  /**
+   * @param {Object} response
+   * @param {Promise} bodyPromise
+   * @param {Object} requestData
+   * @override
+   */
+  ajaxSuccess(response, bodyPromise, requestData) {
+    const headerString = headersToObject(response.headers);
+
+    return bodyPromise.then((body) => {
+      let returnResponse;
+      try {
+        returnResponse = this.handleResponse(
+          response.status,
+          headersToObject(headerString),
+          body,
+          requestData
+        );
+      } catch (error) {
```

Comment:
this `try/catch` is redundant and can be removed. the caught error will already be caught be the `then` handler it is being called in.